src/bin/miri.rs

```diff
@@ -82,6 +84,15 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
                 env::set_current_dir(cwd).unwrap();
             }
 
+            if tcx.sess.opts.optimize != OptLevel::No {
+                tcx.sess.warn("Miri does not support optimizations.");
+            } else if tcx.sess.mir_opt_level() > 0 {
+                tcx.sess.warn("You have explicitly enabled MIR optimizations, overriding Miri's default \
+                    which is to completely disable them. Any optimizations may hide UB that Miri would \
+                    otherwise detect, and it is not necessarily possible to predict what kind of UB will \
+                    be missed.");
+            }
```

Comment:
So if the user says `-O -Zmir-opt-level=3`, they will get the first warning, not the second? That seems wrong?

Reply:
That was definitely wrong.